pyuvdata/uvbeam.py

```diff
@@ -436,6 +442,89 @@ def set_phased_array(self):
+            keep_basis_vector: If True, keep the directionality information and
+                just multiply the efields for each basis vector separately
+                (caution: this is not what is standardly meant by the power beam).
+                Default is False.
+            run_check: Option to check for the existence and proper shapes of
+                required parameters after converting to power. Default is True.
+            check_extra: Option to check optional parameters as well as
+                required ones. Default is True.
+            run_check_acceptability: Option to check acceptable range of the values of
+                required parameters after combining objects. Default is True.
+        """
+        if self.beam_type is not 'efield':
+            raise ValueError('beam_type must be efield')
+
+        efield_data = self.data_array
+        efield_naxes_vec = self.Naxes_vec
+
+        feed_pol_order = [(0, 0)]
+        if self.Nfeeds > 1:
+            feed_pol_order.append((1, 1))
+
+        if calc_cross_pols:
+            self.Npols = self.Nfeeds ** 2
+            if self.Nfeeds > 1:
+                feed_pol_order.extend([(0, 1), (1, 0)])
+        else:
+            self.Npols = self.Nfeeds
+
+        pol_strings = []
+        for pair in feed_pol_order:
+            pol_strings.append(self.feed_array[pair[0]] + self.feed_array[pair[1]])
+        self.polarization_array = np.array([uvutils.polstr2num(ps.upper()) for ps in pol_strings])
+
+        if not keep_basis_vector:
+            self.Naxes_vec = 1
+
+        # adjust requirements, fix data_array form
+        self.set_power()
+        power_data = np.zeros(self._data_array.expected_shape(self), dtype=np.complex)
+
+        if keep_basis_vector:
+            for pol_i, pair in enumerate(feed_pol_order):
+                power_data[:, :, pol_i] = (efield_data[:, :, pair[0]] *
+                                           np.conj(efield_data[:, :, pair[1]]))
+
+        else:
+            for pol_i, pair in enumerate(feed_pol_order):
+                if efield_naxes_vec == 2:
+                    for comp_i in range(2):
+                        comp = ((efield_data[0, :, pair[0]] *
+                                 np.conj(efield_data[0, :, pair[1]])) *
+                                self.basis_vector_array[0, comp_i]**2 +
+                                (efield_data[1, :, pair[0]] *
+                                 np.conj(efield_data[1, :, pair[1]])) *
+                                self.basis_vector_array[1, comp_i]**2 +
+                                (efield_data[0, :, pair[0]] *
+                                 np.conj(efield_data[1, :, pair[1]]) +
+                                 efield_data[1, :, pair[0]] *
+                                 np.conj(efield_data[0, :, pair[1]])) *
+                                (self.basis_vector_array[0, comp_i] *
+                                 self.basis_vector_array[1, comp_i]))
```

Comment:
Oh my. I wrote this out, and the best I can tell it looks right.

Reply:
yeah, it took me several pieces of paper. I *think* it's right and the testing helped me trust it because they were failing until I got it right.